Add guideline to write tests (#185)
DioxusLabs/taffy: CONTRIBUTING.md

To add a new test case add another HTML file to `/test_fixtures` following the current tests as a template for new tests.

### Writing tests

> 1. All tests should be wrapped in a module called `tests`, to ensure they are not compiled unless testing

Reviewer: ```suggestion
1. All tests should be wrapped in a module called `tests` gated by the standard `test` feature flag, to ensure they are not compiled unless tests are being run.

```

    ```rs
    #[cfg(test)]
    mod tests {
        // Place tests here
    }
    ```

> 2. For unit-testing this should be placed within the file of the definition they are testing

Reviewer: ```suggestion
2. For unit-testing this should be placed in the same file as the code they are testing
```

>     1. If file becomes significantly long, tests should be split out into its own file under the same module

Reviewer: ```suggestion
    1. If files become extremely long (more than a couple thousand lines of code), tests should be split out into its own file under the same module
```

    ```rs
    // file: ./src/my_struct.rs
    struct MyStruct;

    impl MyStruct {
        fn some_method() { .. }
    }

    #[cfg(test)]
    mod tests {
        #[test]
        fn test_of_some_method() { .. }
    }
    ```

3. For integration tests this should be placed within the `./tests` folder

    ```rs
    // file: ./tests/my_integration_test.rs
    #[test]
    fn integration_test_one() { .. }
    #[test]
    fn integration_test_two() { .. }
    ```

4. Each test should have a clear intent
    1. It should be evident what is being tested (naming, code, comments)
    2. When this test fails, it should be easy to understand what went wrong

>     3. Ue fixture or parameter-based testing (`#[rstest]`) wherever it improves the two previous points compared to normal rust tests

Reviewer: ```suggestion
    3. Fixture-based tests (`#[rstest]`) can help improve clarity when many related scenarios need to be checked.
```

To add a new test case add another HTML file to `/test_fixtures` following the current tests as a template for new tests.

### Writing tests

1. All tests should be wrapped in a module called `tests`, to ensure they are not compiled unless testing

```rs
#[cfg(test)]
mod tests {
    // Place tests here
}

> ```

Reviewer 2: I'd recommend to indent these with 4 spaces, otherwise some Mardown renderers will terminate the enumeration here and the `2.` will show up as `1.` again.

The same applies for the other items.

2. For unit-testing this should be placed within the file of the definition they are testing
    1. If file becomes significantly long, tests should be split out into its own file under the same module

```rs
// file: ./src/my_struct.rs
struct MyStruct;

impl MyStruct {
    fn some_method() { .. }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_of_some_method() { .. }
}
```

> 4. For integration tests this should be placed within the `./tests` folder

Reviewer 2: Step 3 seems to be missing

```rs
// file: ./tests/my_integration_test.rs
#[cfg(test)]

> mod tests {

Reviewer 2: Is it necessary to wrap this in a module again, when it's in a separate file already?
It might also work to use `#![cfg(test)]` at the top of the file instead, but I haven't tested this out.

Author: IIRC rust treats a file as a module, in which case wrapping it another module is completely redundant. Good spot :)

Author: According to the [rust docs](https://doc.rust-lang.org/book/ch11-03-test-organization.html#the-tests-directory) everything in the `tests` directory will only be compiled on tests, which means we don't need to specify `#![cfg(tests)]`

    #[test]
    fn integration_test_one() { .. }
    #[test]
    fn integration_test_two() { .. }
}
```

5. Each test should have a clear intent
    1. It should be evident what is being tested (naming, code, comments)

>     1. When this test fails, it should be easy to understand why

Author: Markdown will render this correctly as "a numbered list" with correct numbers but it does not hurt to just make sure the source stays clean as well
